### DISCUSSION TITLE
Highlight Signals

### DISCUSSION THREAD

**Discussion by huserben** [nbrown02/ADO-Process-Behaviour-Chart#1] [Ideas]
Hi @nbrown02 

I was wondering if there would be a way to visualize signals in the respective charts. I played a bit around and found a way using this [Youtube Video](https://www.youtube.com/watch?v=A4iG8S4EpHs) and the following measure:

```
WIP_Has_Signal = 
VAR above_UNPL = CALCULATE(MAX(WIP2[Count])) > [UNPL]
VAR below_LNPL = CALCULATE(MIN(WIP2[Count])) < [LNPL]

RETURN IF (above_UNPL || below_LNPL, "!", "")
```

Then I could create a formatting rule for the color (while it's a bar chart), and display red/blue depending on the value (for some reason it only worked when it was a string, not with boolean values...).

The result looks like this:
![SignalsHighlighted](https://github.com/nbrown02/Process-Behaviour-Chart/assets/5486874/b33fe864-5c5f-49c1-b9ce-de37601c9244)

Now the "outliers" above/below the natural process limits are easy to spot, but perhaps it could also be tweaked to find runs of eight above/below the average (probably a bit harder to spot...). Given that it's a "range", I think we could check the previous 7 values and see if they were "on the same side" of the average and just color this dot...

I don't think this is a must, but I could imagine it could be useful to quickly spot signals.

PS: I'm just doing trial and error/Googling, so if there is a better way that you could point me towards, I'm grateful :-)

Cheers!
Benji

**Comment by nbrown02** [comment]
Love it! Feel free to submit a PR, better you claim credit for it :)Sent from my iPhoneOn 1 Jan 2024, at 18:43, Benj Huser ***@***.***> wrote:﻿
Hi @nbrown02
I was wondering if there would be a way to visualize signals in the respective charts. I played a bit around and found a way using this Youtube Video and the following measure:
WIP_Has_Signal = 
VAR above_UNPL = CALCULATE(MAX(WIP2[Count])) > [UNPL]
VAR below_LNPL = CALCULATE(MIN(WIP2[Count])) < [LNPL]

RETURN IF (above_UNPL || below_LNPL, "!", "")

Then I could create a formatting rule for the color (while it's a bar chart), and display red/blue depending on the value (for some reason it only worked when it was a string, not with boolean values...).
The result looks like this:
image.png (view on web)
Now the "outliers" above/below the natural process limits are easy to spot, but perhaps it could also be tweaked to find runs of eight above/below the average (probably a bit harder to spot...). Given that it's a "range", I think we could check the previous 7 values and see if they were "on the same side" of the average and just color this dot...
I don't think this is a must, but I could imagine it could be useful to quickly spot signals.
PS: I'm just doing trial and error/Googling, so if there is a better way that you could point me towards, I'm grateful :-)
Cheers!
Benji

**Reply by huserben** [reply]
Great, I'll have to do some research and clean it up, and then will create a PR.

Btw, I just managed to highlight a "run of eight" using the following custom column and the similar coloring approach:

```
Is_Moderate_Signal = 
IF (
   COUNTROWS (
      FILTER (
         ALL ( CycleTime2 ),
         CycleTime2[Index] >= EARLIER(CycleTime2[Index]) - 7 &&
         CycleTime2[Index] <= EARLIER(CycleTime2[Index]) &&
         CycleTime2[CycleTimeDays] >= CALCULATE([CT_Average])
      )
   ) = 8,
   "Moderate Signal",
   IF (
      COUNTROWS (
         FILTER (
            ALL ( CycleTime2 ),
            CycleTime2[Index] >= EARLIER(CycleTime2[Index]) - 7 &&
            CycleTime2[Index] <= EARLIER(CycleTime2[Index]) &&
            CycleTime2[CycleTimeDays] <= CALCULATE([CT_Average])
         )
      ) = 8,
      "Moderate Signal",
      ""
   )
)
```

![RunOfEight](https://github.com/nbrown02/Process-Behaviour-Chart/assets/5486874/c591f65e-b5b1-4933-a0f5-2eaa924a6a29)

I'll reach out once it's in a "showable" state and create the PR 😉

**Comment by nbrown02** [comment]
Amazing - look forward to seeing it!

**Reply by huserben** [reply]
PR created - see #2 
Propose to close this discussion.

**Reply by nbrown02** [reply]
Thank you 🙏🏼 love it

**Reply by nbrown02** [reply]
@huserben - I took out the marker for the run of 8, after running a few teams' data through it I found it too much. Updated in latest commit :)

**Comment by nbrown02** [comment]
@huserben I was thinking about how to make this better and to incorporate the sigma values/signals but need your help :)

I want to add a selection to filter the type of signals you want to see, so far I have this:

![PBCGIF](https://github.com/nbrown02/Process-Behaviour-Chart/assets/29369962/68ba373f-0ab2-4a03-87e0-c0bf164c21f8)

I wondered if you still had your measures for the sigmas AND if we could amend this formula:

WIA_Is_Signal = 

VAR currentTotalWIAge = SUMX(WIAge2, WIAge2[TotalWIAge])
VAR higher_than_unlp = currentTotalWIAge > [WIAge_UNPL]
VAR lower_than_nlp = currentTotalWIAge < [WIAge_LNPL]
VAR WIAge_AVG = [WIAge_Average]

RETURN
IF (
    higher_than_unlp || lower_than_nlp,
    1,
    IF (
        CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL ( WIAge2 ),
                    WIAge2[Index] >= MIN(WIAge2[Index]) - 7 &&
                    WIAge2[Index] <= MIN(WIAge2[Index]) &&
                    WIAge2[TotalWIAge] > WIAge_AVG
                )
            )
        ) = 8,
        2,
        IF (
            CALCULATE(
                COUNTROWS (
                    FILTER (
                        ALL ( WIAge2 ),
                        WIAge2[Index] >= MIN(WIAge2[Index]) - 7 &&
                        WIAge2[Index] <= MIN(WIAge2[Index]) &&
                        WIAge2[TotalWIAge] < WIAge_AVG
                    )
                )
            ) = 8,
            2,
            0
        )
    )
)

To give a value of 2 for Moderate Change, 3 for Moderate Shift, 4 for Small Shift ?

**Reply by huserben** [reply]
Hey @nbrown02 

I like the idea. That gives the option while keeping the views "clean" :)

I can have a look at it sometime this week (I'm back at work after a 3-week vacation, so there's some catching up to do). Do you want me to provide the measures here? The other is that you provide me with your current template and I directly adjust it...but I think I can create the measures without the "UI" ready...if only Power BI would support  Unit Testing I could do it TDD 😄 

To make sure I understood you correctly:
- "Large": A single point outside the computed natural process limits --> 1
- "Moderate Change": Two out of three successive values that are beyond one of the two sigma lines  --> 2
- "Moderate Shift": Four out of five successive values that are beyond one of the one sigma lines  --> 3
- "Small Shift": Eight successive values on the same side of the average --> 4

If yes, I can deliver the measures with an 85% certainty by end of this week, with a 50% certainty by Wednesday 😄

**Reply by nbrown02** [reply]
@huserben thank you! so I can share with you a PBIT file as all the visuals/bookmarks for the different views are built, then you can just tweak the measures :) 

Will tidy it up tomorrow and share in this thread!

**Comment by nbrown02** [comment]
@huserben If you download from [here](https://drive.google.com/file/d/1kYsJtW18TqveVYTL70yipmiII662ftgL/view?usp=sharing) this is the version I'm playing around with. Changes I've made in this version:
- Added the radio buttons for viewing the types of signals
- Setup bookmarks to navigate between the charts showing different signals
- Improve the readability of the y-axis so markers for values at the top of the chart can be fully seen (previously being cut off)
- Conditional formatting is setup for the different signals based on the above (your understanding is correct)

So the only thing that needs updating are the measures of CT_Is_Signal, TP_Is_Signal, WIA_Is_Signal and WIP_Is_Signal - I'm not fussed about the CT_R_Is_Signal formulas as I only ever want those to highlight when above the URL.

I haven't setup the signals for the PBC (XmR Chart) yet - figured get this page working and release then work on that next :)

**Reply by huserben** [reply]
![Moderate Changes](https://github.com/nbrown02/Process-Behaviour-Chart/assets/5486874/cb77a20c-0e1e-4531-8da1-40b843b61522)

Tried it with the following measures and with a very brief check it worked. One first glance I was confused because a point that is a large change might also be a moderate change or any kind of shift. However, as we define it "per point" they can only have one state, so even though a large change might be also part of a moderate change, it won't be indicated (as the measure already returned 1).
In the above picture you can see this behaviour in the WIP chart.

I don't know Power BI enough to know whether we could do some kind of more complex datatype. But I think it's good enough for a first version, so I'd be ok with how it works now 😉 

## Measures
```
CT_Is_Signal = 

VAR currentTotalCT = SUMX(CycleTime2, CycleTime2[CycleTimeDays])
VAR ct_avg = ROUND([CT_Average], 0)
VAR CT_L_1Sigma = ROUND ( [CT_Average] - (2.66 * [CT_R-mean]) * 0.33, 0 )
VAR CT_L_2Sigma = ROUND ( [CT_Average] - (2.66 * [CT_R-mean]) * 0.66, 0 )
VAR CT_U_1Sigma = ROUND ( [CT_Average] + (2.66 * [CT_R-mean]) * 0.33, 0 )
VAR CT_U_2Sigma = ROUND ( [CT_Average] + (2.66 * [CT_R-mean]) * 0.66, 0 )

VAR large_change = currentTotalCT > [CT_UNPL] || currentTotalCT < [CT_LNPL]

VAR moderate_change = CALCULATE(
            COUNTROWS (
                FILTER (
         ALL   ( CycleTime2 ),
         CycleTime2[Index] >= MIN(CycleTime2[Index]) - 2 &&
         CycleTime2[Index] <= MIN(CycleTime2[Index]) &&
         CycleTime2[CycleTimeDays] > CT_U_2Sigma
      ))
   ) >= 2 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
         ALL   ( CycleTime2 ),
            CycleTime2[Index] >= MIN(CycleTime2[Index]) - 2 &&
            CycleTime2[Index] <= MIN(CycleTime2[Index]) &&
            CycleTime2[CycleTimeDays] < CT_L_2Sigma
         ))
      ) >= 2

VAR moderate_shift = CALCULATE(
            COUNTROWS (
                FILTER (
         ALL   ( CycleTime2 ),
         CycleTime2[Index] >= MIN(CycleTime2[Index]) - 4 &&
         CycleTime2[Index] <= MIN(CycleTime2[Index]) &&
         CycleTime2[CycleTimeDays] > CT_U_1Sigma
      ))
   ) >= 4 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
         ALL   ( CycleTime2 ),
            CycleTime2[Index] >= MIN(CycleTime2[Index]) - 4 &&
            CycleTime2[Index] <= MIN(CycleTime2[Index]) &&
            CycleTime2[CycleTimeDays] < CT_L_1Sigma
         ))
      ) >= 4

VAR small_shift = CALCULATE(
            COUNTROWS (
                FILTER (
         ALL   ( CycleTime2 ),
         CycleTime2[Index] >= MIN(CycleTime2[Index]) - 7 &&
         CycleTime2[Index] <= MIN(CycleTime2[Index]) &&
         CycleTime2[CycleTimeDays] > ct_avg
      ))
   ) = 8 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
         ALL   ( CycleTime2 ),
            CycleTime2[Index] >= MIN(CycleTime2[Index]) - 7 &&
            CycleTime2[Index] <= MIN(CycleTime2[Index]) &&
            CycleTime2[CycleTimeDays] < ct_avg
         ))
      ) = 8


RETURN
IF (large_change, 1,
    IF (moderate_change, 2,
      IF (moderate_shift, 3,
            IF(small_shift, 4, 0)
        )
    )
)
```

```
TP_Is_Signal = 

VAR currentTotalTP = SUMX(Throughput2, Throughput2[Count])
VAR TP_AVG =  ROUND([TP_Average], 0)
VAR TP_L_1Sigma = ROUND ( [TP_Average] - (2.66 * [TP_R-mean]) * 0.33, 0 )
VAR TP_L_2Sigma = ROUND ( [TP_Average] - (2.66 * [TP_R-mean]) * 0.66, 0 )
VAR TP_U_1Sigma = ROUND ( [TP_Average] + (2.66 * [TP_R-mean]) * 0.33, 0 )
VAR TP_U_2Sigma = ROUND ( [TP_Average] + (2.66 * [TP_R-mean]) * 0.66, 0 )

VAR large_change = currentTotalTP > [TP_UNPL] || currentTotalTP < [TP_LNPL]

VAR moderate_change = CALCULATE(
            COUNTROWS (
                FILTER (
         ALL  ( Throughput2 ),
         Throughput2[Index] >= MIN(Throughput2[Index]) - 2 &&
         Throughput2[Index] <= MIN(Throughput2[Index]) &&
         Throughput2[Count] > TP_U_2Sigma
      ))
   ) >= 2 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
         ALL  ( Throughput2 ),
            Throughput2[Index] >= MIN(Throughput2[Index]) - 2 &&
         Throughput2[Index] <= MIN(Throughput2[Index]) &&
            Throughput2[Count] < TP_L_2Sigma
         ))
      ) >= 2

VAR moderate_shift = CALCULATE(
            COUNTROWS (
                FILTER (
         ALL  ( Throughput2 ),
         Throughput2[Index] >= MIN(Throughput2[Index]) - 4 &&
         Throughput2[Index] <= MIN(Throughput2[Index]) &&
         Throughput2[Count] > TP_U_1Sigma
      ))
   ) >= 4 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
         ALL  ( Throughput2 ),
            Throughput2[Index] >= MIN(Throughput2[Index]) - 4 &&
         Throughput2[Index] <= MIN(Throughput2[Index]) &&
            Throughput2[Count] < TP_L_1Sigma
         ))
      ) >= 4

VAR small_shift = CALCULATE(
            COUNTROWS (
                FILTER (
         ALL  ( Throughput2 ),
         Throughput2[Index] >= MIN(Throughput2[Index]) - 7 &&
         Throughput2[Index] <= MIN(Throughput2[Index]) &&
         Throughput2[Count] > TP_AVG
      ))
   ) = 8 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
         ALL  ( Throughput2 ),
            Throughput2[Index] >= MIN(Throughput2[Index]) - 7 &&
         Throughput2[Index] <= MIN(Throughput2[Index]) &&
            Throughput2[Count] < TP_AVG
         ))
      ) = 8


RETURN
IF (large_change, 1,
    IF (moderate_change, 2,
      IF (moderate_shift, 3,
            IF(small_shift, 4, 0)
        )
    )
)
```

```
WIA_Is_Signal = 

VAR currentTotalWIAge = SUMX(WIAge2, WIAge2[TotalWIAge])
VAR WIAge_AVG = [WIAge_Average]
VAR WIAge_L_1Sigma = ROUND ( [WIAge_Average] - (2.66 * [WIAge_R-mean]) * 0.33, 0 )
VAR WIAge_L_2Sigma = ROUND ( [WIAge_Average] - (2.66 * [WIAge_R-mean]) * 0.66, 0 )
VAR WIAge_U_1Sigma = ROUND ( [WIAge_Average] + (2.66 * [WIAge_R-mean]) * 0.33, 0 )
VAR WIAge_U_2Sigma = ROUND ( [WIAge_Average] + (2.66 * [WIAge_R-mean]) * 0.66, 0 )

VAR large_change = currentTotalWIAge > [WIAge_UNPL] || currentTotalWIAge < [WIAge_LNPL]

VAR moderate_change = CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL ( WIAge2 ),
                    WIAge2[Index] >= MIN(WIAge2[Index]) - 2 &&
                    WIAge2[Index] <= MIN(WIAge2[Index]) &&
                    WIAge2[TotalWIAge] > WIAge_U_2Sigma
                )
            )
        ) >= 2 || 
      CALCULATE(
                COUNTROWS (
                    FILTER (
                        ALL ( WIAge2 ),
                        WIAge2[Index] >= MIN(WIAge2[Index]) - 2 &&
                        WIAge2[Index] <= MIN(WIAge2[Index]) &&
                        WIAge2[TotalWIAge] < WIAge_L_2Sigma
                    )
                )
            ) >= 2

VAR moderate_shift = CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL ( WIAge2 ),
                    WIAge2[Index] >= MIN(WIAge2[Index]) - 4 &&
                    WIAge2[Index] <= MIN(WIAge2[Index]) &&
                    WIAge2[TotalWIAge] > WIAge_U_1Sigma
                )
            )
        ) >= 4 || 
      CALCULATE(
                COUNTROWS (
                    FILTER (
                        ALL ( WIAge2 ),
                        WIAge2[Index] >= MIN(WIAge2[Index]) - 4 &&
                        WIAge2[Index] <= MIN(WIAge2[Index]) &&
                        WIAge2[TotalWIAge] < WIAge_L_1Sigma
                    )
                )
            ) >= 4

VAR small_shift = CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL ( WIAge2 ),
                    WIAge2[Index] >= MIN(WIAge2[Index]) - 7 &&
                    WIAge2[Index] <= MIN(WIAge2[Index]) &&
                    WIAge2[TotalWIAge] > WIAge_AVG
                )
            )
        ) = 8 || 
      CALCULATE(
                COUNTROWS (
                    FILTER (
                        ALL ( WIAge2 ),
                        WIAge2[Index] >= MIN(WIAge2[Index]) - 7 &&
                        WIAge2[Index] <= MIN(WIAge2[Index]) &&
                        WIAge2[TotalWIAge] < WIAge_AVG
                    )
                )
            ) = 8


RETURN
IF (large_change, 1,
    IF (moderate_change, 2,
      IF (moderate_shift, 3,
            IF(small_shift, 4, 0)
        )
    )
)
```

```
WIP_Is_Signal = 

VAR currentTotalWIP = SUMX(WIP2, WIP2[Count])
VAR WIP_avg = ROUND([WIP_Average], 0)

VAR WIP_L_1Sigma = ROUND ( [WIP_Average] - (2.66 * [WIP_R-mean]) * 0.33, 0 )
VAR WIP_L_2Sigma = ROUND ( [WIP_Average] - (2.66 * [WIP_R-mean]) * 0.66, 0 )
VAR WIP_U_1Sigma = ROUND ( [WIP_Average] + (2.66 * [WIP_R-mean]) * 0.33, 0 )
VAR WIP_U_2Sigma = ROUND ( [WIP_Average] + (2.66 * [WIP_R-mean]) * 0.66, 0 )

VAR large_change = currentTotalWIP> [WIP_UNPL] || currentTotalWIP < [WIP_LNPL]

VAR moderate_change = CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL ( WIP2 ),
                        WIP2[Index] >= MIN(WIP2[Index]) - 2 &&
                        WIP2[Index] <= MIN(WIP2[Index]) &&
                        WIP2[Count] > WIP_U_2Sigma
      ))
   ) >= 2 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL  ( WIP2 ),
                    WIP2[Index] >= MIN(WIP2[Index]) - 2 &&
                    WIP2[Index] <= MIN(WIP2[Index]) &&
                    WIP2[Count] < WIP_L_2Sigma
                    )
                )
        ) >= 2


VAR moderate_shift = CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL ( WIP2 ),
                        WIP2[Index] >= MIN(WIP2[Index]) - 4 &&
                        WIP2[Index] <= MIN(WIP2[Index]) &&
                        WIP2[Count] > WIP_U_1Sigma
      ))
   ) >= 4 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL  ( WIP2 ),
                    WIP2[Index] >= MIN(WIP2[Index]) - 4 &&
                    WIP2[Index] <= MIN(WIP2[Index]) &&
                    WIP2[Count] < WIP_L_1Sigma
                    )
                )
        ) >= 4


VAR small_shift = CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL ( WIP2 ),
                        WIP2[Index] >= MIN(WIP2[Index]) - 7 &&
                        WIP2[Index] <= MIN(WIP2[Index]) &&
                        WIP2[Count] > WIP_avg
      ))
   ) = 8 || 
      CALCULATE(
            COUNTROWS (
                FILTER (
                    ALL  ( WIP2 ),
                    WIP2[Index] >= MIN(WIP2[Index]) - 7 &&
                    WIP2[Index] <= MIN(WIP2[Index]) &&
                    WIP2[Count] < WIP_avg
                    )
                )
        ) = 8


RETURN
IF (large_change, 1,
    IF (moderate_change, 2,
      IF (moderate_shift, 3,
            IF(small_shift, 4, 0)
        )
    )
)
```

**Comment by nbrown02** [comment]
Amazing thank you so much - I've run them through Dax Formatter too to
neaten them up a bit (along with all the other measures). Going to add
these to the XmR page then push live this morning :)

**Comment by nbrown02** [comment]
@huserben all done in latest version published :)

**Reply by huserben** [reply]
Awesome, will try it out as soon as find some time!
